### DISCUSSION TITLE
API to automate push and pull for sites

### DISCUSSION THREAD

**Discussion by metmarkosaric** [plausible/analytics#150] [Anything Else]
"How about an API to push and pull data from a site you own?

    Manage sites
    Generate script to insert in site HEAD
    Retrieve information by /API/myUserID/mySite/{Method}

Output:

    JSON formatted to parse it.
    Happyness

I would also like an API to manage sites in plausible, create, update and delete. If an API is exposed for this I will write a terraform provider for it."

([imported from the old roadmap](https://plausible.nolt.io/111))

**Comment by kurtmc** [comment]
I had commented on the old roadmap asking for create/update/delete API for sites as I am interested in implementing a terraform provider for plausible so that I can manage my sites using it.

**Comment by mrcave** [comment]
WP Multisite provider here. Would love to power some stats for our 700+ sites with plausible but not feasible without an API.

**Comment by ritvvijparrikh** [comment]
+1

1. APIs to create, delete sites
2. API to fetch analytics for whole site (cumulative or for a given data range)
3. API to fetch analytics for one page (cumulative or for a given data range)

**Comment by kurtmc** [comment]
I decided that I can work on the terraform provider even without an official API, if an official API gets supported in the future I will switch over to using that.

If anyone is interested in following along, I have created a terraform provider for plausible but it's still early days and I have only implemented creating and deleting of the site resource, hopefully we will get more features implemented over the next few months.

https://github.com/mcalpinefree/terraform-provider-plausible

**Comment by kmahelona** [comment]
Would love an API so we can pull in the analytics into a custom dashboard. We also serve content on the web and on mobile and being able to POST custom events from a mobile app would be great so we don't have to run separate analytics platforms for web and mobile.

**Comment by metmarkosaric** [comment]
We plan to work on the API as the next big project at the start of next year. In order to deliver the product that satisfies your needs, we would love to hear from people who want to offer Plausible Analytics to their customers, clients etc using this API.

What do you plan do to? What do you need and expect from the API? Do you have any specific requirements that we should consider when building this? 

You can share your feedback here or reach out to us on hello@plausible.io. Thanks!

**Reply by simonalbrecht** [reply]
We’re having the use case that we’d like to use Plausible in a SaaS product that allows user to create their own web presence/website on a subdomain.

For this to work, we’d need the API to be able to automate site creation, modification and deletion. 

For example if a user signs up in our product, we want to automatically create a new site in Plausible that only contains the data of that user. 

With this in place, each users analytics data would independent from the data of other users.

**Reply by ukutaht** [reply]
Makes sense. This sort of use-case is the main one we're targeting with this API addition.

It will be available on our managed SaaS first. If you're interested in using our managed offering, please write to us on the email above what sort of volume you'd be bringing. We'll let you know once the API is available.

**Reply by binaryfire** [reply]
@ukutaht @metmarkosaric My use case is the same as Simon. I'm developing a website builder for musicians and need to programatically add / remove sites when a customer signs up / cancels. It's all DIY and using open source tools, so I'm waiting for this to come to the self hosted version :)

I'm currently using Matomo and this is the only thing stopping me from switching to Plausible :)

**Comment by ArsalaBangash** [comment]
Eagerly awaiting the create API to help automate generating plausible analytics for a large number of sites 😃

**Comment by jivinivan** [comment]
+1 for an api to add and remove sites

**Comment by petethered** [comment]
+1 for an API to add sites

**Comment by metmarkosaric** [comment]
The site provisioning API is now in private preview. It's a work in progress but currently it allows 2 basic operations:

* Create a new site
* Find or create a shared link by name

You can learn more here https://plausible.io/docs/sites-api

**Reply by mattvb91** [reply]
Is there any ETA when this is available on self hosted?

**Comment by RobertJoonas** [comment]
Our site Provisioning API now also supports deleting existing sites from Plausible. Take a look here for the documentation: https://plausible.io/docs/sites-api#delete-apiv1sitessite_id.
